### PR TITLE
Porting guide: Add Sleep API pg page

### PR DESCRIPTION
**Pull request by bulislaw**
Status: *Work in progress*

It's basically Martin's proposal.

@c1728p9 @0xc0170

### CODE REVIEW

File: docs/contributing/sleep_deep_sleep.md [hunk 1]
```diff
@@ -1,3 +1,28 @@
...
+1. Latency - should wake-up within 10 us
+
+## Deep sleep
+
+The core system clock is disabled. The low precision clocks are enabled, RAM is retained.
+
+1. Wake-up sources - RTC, low power ticker or GPIO should be able to wake-up the MCU
+1. Latency - should wake-up within 5 ms
+
+The deep sleep latency (5 ms) was chosen as the higher limit of the vendors we support. Most of targets have wake-up latency for deep sleep within few microseconds. This wake-up time does not include reinitializing clocks or other related configuration to return to previous state after the deep sleep.
```
Comment (bulislaw) [inline docs/contributing/sleep_deep_sleep.md]: I'm not sure I like that. We should either give a strict limit to go back to previous state or explicitly mark it as undefined.

Comment (0xc0170) [reply docs/contributing/sleep_deep_sleep.md]: I believe our intention was to have those 10ms (we increased it from 5) as a strict limit. Most of MCU can wake up within few microseconds, the part that takes are clocks related config reinit, up to few milliseconds. Shall we update this accordingly ?

File: docs/contributing/sleep_deep_sleep.md [hunk 2]
```diff
@@ -1,3 +1,28 @@
...
+
+1. Wake-up sources - any interrupt source should be able to wake-up the MCU
+1. Latency - should wake-up within 10 us
+
+## Deep sleep
+
+The core system clock is disabled. The low precision clocks are enabled, RAM is retained.
+
+1. Wake-up sources - RTC, low power ticker or GPIO should be able to wake-up the MCU
+1. Latency - should wake-up within 5 ms
```
Comment (0xc0170) [inline docs/contributing/sleep_deep_sleep.md]: 10ms

File: docs/contributing/sleep_deep_sleep.md [hunk 3]
```diff
@@ -1,3 +1,52 @@
...
+The core system clock is disabled. The low precision clocks are enabled, RAM is retained.
+
+1. Wake-up sources - RTC, low power ticker or GPIO should be able to wake-up the MCU
+1. Latency - should wake-up within 10 ms
+
+The deep sleep latency (10 ms) was chosen as the higher limit of the boards we support. Most of targets have wake-up latency for deep sleep within few microseconds, but often additional time is needed for reinitializing clocks or other configuration necessary to restore previous state.
+
+## Implementing the Sleep API
+
+There are two functions that target needs to implement to support sleep, their prototypes are located in [hal/sleep_api.h]():
```
Comment (bulislaw) [inline docs/contributing/sleep_deep_sleep.md]: We'll embed the header here instead of copying and linking, when the changes are committed.

**Comment by AnotherButler** [comment]
@bulislaw I've added some minor copy edits. Please review, and make sure I didn't accidentally change the content.

**Comment by bulislaw** [comment]
Thanks, looks good :)

**Comment by AnotherButler** [comment]
@bulislaw Is this ready to merge, or is is still waiting on any engineering review/changes?

**Comment by bulislaw** [comment]
I think I've addressed all the points raised by @0xc0170 so we should be ready to merge.

File: docs/contributing/sleep_deep_sleep.md [hunk 4]
```diff
@@ -1,3 +1,52 @@
...
+
+```c++
+void hal_deepsleep(void);
+```
+
+To enable sleep support in Mbed OS, you need to add the `SLEEP` label in the `device_has` option of the target's section in the `targets.json` file.
+
+### Testing
+
+The [sleep HAL API test suite]() validates:
```
Comment (AnotherButler) [inline docs/contributing/sleep_deep_sleep.md]: Query: Where should this link?

Comment (AnotherButler) [reply docs/contributing/sleep_deep_sleep.md]: @0xc0170 Do you know where this should link?

Comment (AnotherButler) [reply docs/contributing/sleep_deep_sleep.md]: @c1728p9 Do you know where this should link?

Comment (c1728p9) [reply docs/contributing/sleep_deep_sleep.md]: Lets remove the testing for now and add it in a seperate PR once there is a link.

**Pull request merged**